sway/tree/arrange.c

```diff
@@ -71,16 +71,92 @@ void arrange_workspace(struct sway_container *workspace) {
 	struct wlr_box *area = &output->sway_output->usable_area;
 	wlr_log(L_DEBUG, "Usable area for ws: %dx%d@%d,%d",
 			area->width, area->height, area->x, area->y);
-	workspace->width = area->width;
-	workspace->height = area->height;
-	workspace->x = area->x;
-	workspace->y = area->y;
+	if (config->edge_gaps || (config->smart_gaps && workspace->children->length > 1)) {
+		double gaps = workspace->has_gaps ? workspace->gaps_outer : config->gaps_outer;
+		workspace->width = area->width - (2 * gaps);
+		workspace->height = area->height - (2 * gaps);
+		workspace->x = area->x + gaps;
+		workspace->y = area->y + gaps;
+	} else {
+		workspace->width = area->width;
+		workspace->height = area->height;
+		workspace->x = area->x;
+		workspace->y = area->y;
+	}
 	wlr_log(L_DEBUG, "Arranging workspace '%s' at %f, %f",
 			workspace->name, workspace->x, workspace->y);
 	arrange_children_of(workspace);
 	container_damage_whole(workspace);
 }
 
+static void remove_gaps(struct sway_container *c) {
+	if (c->gapped) {
+		double gaps = c->current_gaps;
+
+		if (c->parent && c->parent->type == C_CONTAINER) {
+			if (c->parent->children->length > 1) {
+				bool is_first_child = c == c->parent->children->items[0];
+				if (c->parent->layout == L_VERT) {
+					if (!is_first_child) {
+						c->y -= gaps;
+					}
+					c->height += 2 * gaps;
+				} else if (c->parent->layout == L_HORIZ) {
+					if (!is_first_child) {
+						c->x -= gaps;
+					}
+					c->width += gaps;
+				}
+			}
+		} else {
+			c->width += 2 * gaps;
+			c->height += 2 * gaps;
+			c->x -= gaps;
+			c->y -= gaps;
+		}
+
+		c->gapped = false;
+		c->current_gaps = 0;
+		wlr_log(L_DEBUG, "Removing gaps %p", c);
+	} else {
+		wlr_log(L_DEBUG, "Removing gaps: not gapped: %p", c);
+	}
+}
+
+static void add_gaps(struct sway_container *c) {
```

Comment:
The code for `add_gaps` is the same as `remove_gaps`, with s/+/-/. Is it possible to unify these in a `set_gaps` function?

Reply:
I was actually going to change the way gaps were set. Right now, too much of how gaps are removed depends on where the sway_container is. With things like flattening, some ugly bugs are going to pop up. I was going to change it so that a sway_container tracks gaps on each dimension (gaps_width, gaps_y, etc). This will make the remove_gaps() function smaller and more DRY.